### DISCUSSION TITLE
Custom HTTP Headers Amazon S3 File Writer variable not updating after first being set

### DISCUSSION THREAD

**Discussion by KevinHobbsSeattle** [nextgenhealthcare/connect#5024] [Q&A] [answered]
I have an S3 File Writer destination trying to use a ${} variable in the custom HTTP headers and it's getting stuck on the first value in the first message processed through the channel and never updates dynamically to the following messages. The first message is fine, the value is being used from the setting in the source using the channelMap. But after that the following messages, which are changing the value in the channel map, but the destination ${} lookup never changes, it get stucks on the first value it's set to.

<img width="1112" alt="Screen Shot 2022-02-14 at 5 26 48 PM" src="https://user-images.githubusercontent.com/12946276/153975308-a86097b8-6668-4de4-a3bf-6ca5530d1cf6.png">

Mirth 3.12.0
OpenJDK 11.0.2
Ubuntu 16.04

**Comment by jonbartels** [comment]
`com.mirth.connect.connectors.file.FileDispatcher#replaceConnectorProperties` appears to make the right call to substitute the properties on each message:

```
S3SchemeProperties s3Properties = (S3SchemeProperties) schemeProperties;
s3Properties.setRegion(replacer.replaceValues(s3Properties.getRegion(), connectorMessage));
s3Properties.setCustomHeaders(replacer.replaceKeysAndValuesInMap(s3Properties.getCustomHeaders(), connectorMessage));
```
            
I am not an expert with S3. I will let another guru comment about S3 specifically.

When you run your messages through, can you look at the "mappings" tab for each message and assert that:

* `clinic` is being set on a message context map like source, connector, or channel BUT NOT global-channel or global?
* `clinic` does indeed change between messages
* Show, but do not change, your advance queue settings. I wonder if "regen templates" is in play here...

These will rule out a generic MC mapping typo or bug.

**Reply by KevinHobbsSeattle** [reply]
Thank you for your response. I have double checked everything. I'm using the channelMap in the source. The mappings tab shows the value being updated. Looking at the sent tab even shows the value changing for [CUSTOM HEADERS]
x-amz-meta-clinic: <clinic_name>. So it appears mirth is doing the correct thing, it's sending changed values into the AWS S3 API, or is at least telling me that it is. But, when I look on S3 the tag never changes from what it is the first time sent.
Redeploying the channel does make it work again, but only for the first time, and then never changes.

It's as if there is an issue with the AWS API or HTTP HEADERS getting cached somewhere.
Anyway, I'll have to move on and see if using a java api call directly works better.

**Reply by KevinHobbsSeattle** [reply]
The Advanced Queue Settings are set to this
<img width="293" alt="Screen Shot 2022-02-15 at 5 20 46 PM" src="https://user-images.githubusercontent.com/12946276/154178168-f911bacf-ab2f-40f0-a9af-6a377b4666f1.png">
:

**Reply by KevinHobbsSeattle** [reply]
I also tried these and there was no difference
<img width="289" alt="Screen Shot 2022-02-15 at 5 19 48 PM" src="https://user-images.githubusercontent.com/12946276/154178228-05b4175c-2fe4-4828-bb0f-684dbf9d3874.png">
:

**Reply by jonbartels** [reply]
Hmm my money was on Regenerate Templates.

Does the response message from S3 have any information in it? I am a dinosaur and not familiar with S3, are you trying to do something like an update operation on an object that doesn't allow updates? 

I see in the help text in your screenshot that "these headers will be included in all PUT requests", is the operation you're performing a PUT or a POST?

**Comment by ChristopherSchultz** [comment]
Even though it should be obvious, can you post the code/config you use to set that channelMap variable?

**Comment by pacmano1** [comment]
I can reproduce this.   Note the user defined metadata on the second screenshot should be a 6, not a 5.

The first message sent on the poll was "5.txt" and posted with `x-amz-meta-message-id: 5` (and 5 was in the metadata as expected), the second message was "6.txt" as shown in these screen shots.

![image](https://user-images.githubusercontent.com/44065187/154314590-45e1d59e-d5eb-4138-a7a0-f8b15285a22b.png)

![image](https://user-images.githubusercontent.com/44065187/154314649-da3b25db-9476-4b32-8a93-a43fc920fcb3.png)

**Reply by ChristopherSchultz** [reply]
Is this unique to the S3 file writer? What if you switch to an HTTP writer with similar custom headers and point it at a simple "echo" web server?

**Comment by jonbartels** [comment, marked as answer]
An issue was created for this problem https://github.com/nextgenhealthcare/connect/issues/5116
